File: simulations/ExampleIntroduction.R
## Code to reproduce Figure 1
# Introductory example

# The R-Script VarP.R needs to be run before.

# consider the setting n = 300, p = 800 from the simulation script VarP.R
load("SimulationResults/VarP_2024_08_13/equalCI_P_800_rho00.RData")

nrep <- 100
p.vec <- c(50, 100, 200, 400, 800)

df <- data.frame(matrix(NA, ncol = 3, nrow = nrep * 3))
colnames(df) <- c("meth", "MSE", "s.active")
count <- 1
for(l in 1:nrep){
  df[count, ] <- data.frame(meth = "deconfounded", MSE = lres.rhoNULL[[l]][1,1], s.active = lres.rhoNULL[[l]][2,1])
  df[count+1, ] <- data.frame(meth = "naive", MSE = lres.rhoNULL[[l]][1,2], s.active = lres.rhoNULL[[l]][2,2])
  df[count + 2    , ] <- data.frame(meth = "estimated factors", MSE = lres.rhoNULL[[l]][1,3], s.active = lres.rhoNULL[[l]][2,3])
  count <- count + 3
}


pdf("PlotResults/FinalPlots/ExampleIntroduction.pdf", width = 7, height = 4)
par(mfrow = c(1,2))
hist(df[df$meth == "deconfounded", "MSE"], xlim = c(0, 12.5), breaks = seq(0, 12.5, 0.25),
     col = rgb(0,0, 1,0.2), xlab = "MSE", main = "Mean squared error of f")
hist(df[df$meth == "naive", "MSE"], add = TRUE, breaks = seq(0, 12.5, 0.25), col = rgb(1,0, 0,0.2))

legend("topright", legend=c("deconfounded","naive"), col=c(rgb(0,0,1,0.2), 
                                                      rgb(1,0,0,0.2)), pt.cex=2, pch=15 )

hist(df[df$meth == "deconfounded", "s.active"], xlim = c(0, 145), ylim = c(0, 22), breaks = seq(0, 145, 5),
     col = rgb(0,0, 1,0.2), xlab = "Size of estimated active set", main = "Size of estimated active set")
hist(df[df$meth == "naive", "s.active"], add = TRUE, breaks = seq(0, 145, 5), col = rgb(1,0, 0,0.2))
legend("topright", legend=c("deconfounded","naive"), col=c(rgb(0,0,1,0.2), 
                                                   rgb(1,0,0,0.2)), pt.cex=2, pch=15 )
abline(v=4)

dev.off()

File: simulations/example_introduction.py
"""
Reproduce Figure 1.

Introductory example: histograms of MSE and active set size for the
setting n=300, p=800 from the VarP simulation.

Requires var_p.py to have been run first (p=800, rho=NULL, equal CI).

Equivalent to R's SimulationScripts/ExampleIntroduction.R.

Usage
-----
python example_introduction.py
"""

from pathlib import Path
import sys

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from _sim_utils import load_results, PALETTE, METHS

# ---------------------------------------------------------------------------
# Load results for p=800, rho=NULL, equal confounding influence
# ---------------------------------------------------------------------------

results_path = Path("SimulationResults/VarP/P_800_rhoNULL.pkl")

if not results_path.exists():
    raise FileNotFoundError(f"{results_path} not found. Run var_p.py first.")

results = load_results(results_path)
n_rep = min(100, len(results))

mse_dec = [r["MSE"][0] for r in results[:n_rep]]  # deconfounded
mse_naive = [r["MSE"][1] for r in results[:n_rep]]  # naive
act_dec = [r["active"][0] for r in results[:n_rep]]
act_naive = [r["active"][1] for r in results[:n_rep]]

# ---------------------------------------------------------------------------
# Plot (Figure 1)
# ---------------------------------------------------------------------------

PLOT_DIR = Path("PlotResults/FinalPlots")
PLOT_DIR.mkdir(parents=True, exist_ok=True)

fig, axes = plt.subplots(1, 2, figsize=(9, 4))

# --- Left: MSE histogram ---
ax = axes[0]
bins_mse = np.arange(0, 13.0, 0.25)
ax.hist(
    mse_dec,
    bins=bins_mse,
    color=PALETTE["deconfounded"],
    alpha=0.5,
    label="deconfounded",
)
ax.hist(mse_naive, bins=bins_mse, color=PALETTE["naive"], alpha=0.5, label="naive")
ax.set_xlim(0, 12.5)
ax.set_xlabel("MSE", fontsize=11)
ax.set_ylabel("Frequency", fontsize=11)
ax.set_title("Mean squared error of f", fontsize=11)
ax.legend(fontsize=9)

# --- Right: active set size histogram ---
ax = axes[1]
bins_act = np.arange(0, 146, 5)
ax.hist(
    act_dec,
    bins=bins_act,
    color=PALETTE["deconfounded"],
    alpha=0.5,
    label="deconfounded",
)
ax.hist(act_naive, bins=bins_act, color=PALETTE["naive"], alpha=0.5, label="naive")
ax.axvline(x=4, color="black", linewidth=1.0)  # true active set size
ax.set_xlim(0, 145)
ax.set_xlabel("Size of estimated active set", fontsize=11)
ax.set_ylabel("Frequency", fontsize=11)
ax.set_title("Size of estimated active set", fontsize=11)
ax.legend(fontsize=9)

fig.tight_layout()
out_path = PLOT_DIR / "ExampleIntroduction.pdf"
fig.savefig(out_path)
plt.close(fig)
print(f"Saved {out_path}")
